### PR TITLE
Optionally bypass state root verification in reference test worldstate

### LINKED ISSUE #5934: besu evmt8n daemon errors out on test data
out=$(curl -s --data-raw $req 'http://localhost:3000')

**req**
```
{"state":{"fork":"Berlin","reward":"0","chainid":"1"},"input":{"env":{"currentCoinbase":"0x2adc25665018aa1fe0e6bc666dac8fc2697ff9ba","currentNumber":"0x01","currentTimestamp":"0x03e8","currentGasLimit":"0xff112233445566","previousHash":"0x7d46f90648e1c009b83cdec706549090f3676e92fb0ca3cb85de687b73148e13","parentTimestamp":"0x00","parentDifficulty":"0x020000","parentUncleHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","blockHashes":{"0":"0x7d46f90648e1c009b83cdec706549090f3676e92fb0ca3cb85de687b73148e13"}},"alloc":{"0x095e7baea6a6c7c4c2dfeb977efac326af552d87":{"balance":"0x0de0b6b3a7640000","code":"0x600160010160005500","nonce":"0x00","storage":{}},"0x2adc25665018aa1fe0e6bc666dac8fc2697ff9ba":{"balance":"0x00","code":"0x","nonce":"0x01","storage":{}},"0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b":{"balance":"0x0de0b6b3a7640000","code":"0x","nonce":"0x00","storage":{}}},"txs":"0xf865f863800a83061a8094095e7baea6a6c7c4c2dfeb977efac326af552d87830186a0801ba0ffb600e63115a7362e7811894a91d8ba4330e526f22121c994c4692035dfdfd5a06198379fcac8de3dbfac48b165df4bf88e2088f294b61efb9a65fe2281c76e16"}}
```
.retesteth/besu/setup.sh Besu daemon is listening on port 3000

**out**
```
{"error":"World State Root does not match expected value, header 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470 calculated 0xcd39e0cdd18f8f811911222ae6779341663d0293e1a3d9501da7ac2f4da9b277","stacktrace":"java.lang.RuntimeException: World State Root does not match expected value, header 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470 calculated 0xcd39e0cdd18f8f811911222ae6779341663d0293e1a3d9501da7ac2f4da9b277
	at org.hyperledger.besu.ethereum.bonsai.worldview.BonsaiWorldState.persist(BonsaiWorldState.java:398)
	at org.hyperledger.besu.evmtool.T8nExecutor.runTest(T8nExecutor.java:380)
	at org.hyperledger.besu.evmtool.T8nServerSubCommand.handleT8nRequest(T8nServerSubCommand.java:141)
	at org.hyperledger.besu.evmtool.T8nServerSubCommand.handle(T8nServerSubCommand.java:95)
	at org.hyperledger.besu.evmtool.T8nServerSubCommand.lambda$run$0(T8nServerSubCommand.java:78)
	at io.vertx.core.impl.future.FutureImpl$1.onSuccess(FutureImpl.java:91)
	at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:86)
	at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:163)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:51)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)
	at io.vertx.core.http.impl.HttpEventHandler.handleEnd(HttpEventHandler.java:79)
	at io.vertx.core.http.impl.Http1xServerRequest.onEnd(Http1xServerRequest.java:581)
	at io.vertx.core.http.impl.Http1xServerRequest.handleEnd(Http1xServerRequest.java:564)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:76)
	at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:153)
	at io.vertx.core.http.impl.Http1xServerConnection.onEnd(Http1xServerConnection.java:191)
	at io.vertx.core.http.impl.Http1xServerConnection.onContent(Http1xServerConnection.java:181)
	at io.vertx.core.http.impl.Http1xServerConnection.handleOther(Http1xServerConnection.java:161)
	at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:149)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:157)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:109)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
"}
parse error: Invalid string: control characters from U+0000 through U+001F must be escaped at line 56, column 1
parse error: Invalid string: control characters from U+0000 through U+001F must be escaped at line 56, column 1

``

### PR DESCRIPTION
**Pull request by garyschulte**
## PR description
if the passed block header has a stateroot of Hash.ZERO, bypass bonsai state root verification in bonsai reference test worldstate only.

## Fixed Issue(s)

fixes #5934

### CODE REVIEW

**Comment by siladu** [comment]
The is beyond my current understanding of Bonsai, but please can you explain why doing this doesn't undermine the integrity of the tests? IOW, how do we know this won't result in missing some genuine mismatch errors?

**Comment by garyschulte** [comment]
> The is beyond my current understanding of Bonsai, but please can you explain why doing this doesn't undermine the integrity of the tests? IOW, how do we know this won't result in missing some genuine mismatch errors?

Bonsai defers calculating a state root until we 'persist' the worldstate.  When we persist, it is always in the context of a blockheader.  As an additional check to ensure consistency, we validate that the stateroot in the blockheader is the same as the one calculated when persisting the worldstate.

There is a bit of a chicken and egg issue with the reference tests however, because the tests are designed as an output rather than an input.  See Dimitry's [comment](https://github.com/hyperledger/besu/issues/5934#issuecomment-1735218988) - when writing tests something has to provide the stateroot as an output as some point.

As far as not undermining the integrity of the tests, in the original [pr 5686]() we:
* intercept the state test inputs and [rewrite the generated block header](https://github.com/hyperledger/besu/pull/5686/files#diff-177f9c6250a28429fd4d60de9e28c227d4e2d36a8dbefad1f76842053491d748R69-R73) with the correct state root.
* add [currentStateRoot](https://github.com/hyperledger/besu/pull/5686/files#diff-4a1474d043a039440241468b27b762a601d190bf160e3128d52affce27cf48c3R107-R131) to ReferenceTestEnv and all of our 'invented here' test cases

So at least when we run the reference test suites via besu we will be validating the state root. With currentStateRoot, we have the ability to validate stateRoot for any of our own tests like the ones for t8n tool.  All of our current tests except for the one I added in this PR have a non zero stateroot, but we will have to be mindful in the future when developing reference tests to include currentStateRoot.

**Comment by siladu** [comment]
Thanks for the clear explanation!

> but we will have to be mindful in the future when developing reference-test style tests to include currentStateRoot.

This concerns me a bit. How do you think we can mitigate this? I find this area of besu mysterious because it relates to external tools and not well documented so feel like this could easily slip under the radar in the future.

File: ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/BonsaiReferenceTestWorldState.java [hunk 1]
```diff
@@ -75,6 +75,14 @@ public ReferenceTestWorldState copy() {
         layerCopy, cachedMerkleTrieLoader, trieLogManager, preImageProxy);
   }
 
+  @Override
+  protected void verifyWorldStateRoot(final Hash calculatedStateRoot, final BlockHeader header) {
+    // if the supplied header has a non-zero state root, verify.  Else bypass
```
Comment (siladu) [inline ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/BonsaiReferenceTestWorldState.java:80]: nit: would be nice to say why we bypass here

**Review by siladu — APPROVED** [review]

**Comment by garyschulte** [comment]
> Thanks for the clear explanation!
> 
> > but we will have to be mindful in the future when developing reference-test style tests to include currentStateRoot.
> 
> This concerns me a bit. How do you think we can mitigate this? I find this area of besu mysterious because it relates to external tools and not well documented so feel like this could easily slip under the radar in the future.

I suspect that the balance of example will be enough when we develop more of our own reference-style tests.  More often than not, tests cases borrow a lot from existing examples, and all of our examples (save this one) have a non-zero stateroot.

Willing to be convinced otherwise though.

File: ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/BonsaiReferenceTestWorldState.java (hunk 1 above)
Comment (garyschulte) [reply ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/BonsaiReferenceTestWorldState.java]: added a verbose javadoc 👍

**Review by matkt — APPROVED** [review]
